src/kdbindings/connection_evaluator.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <functional>
+#include <list>
+#include <mutex>
+
+namespace KDBindings {
+
+/**
+ * @brief Manages and evaluates conditions for deferred connections.
+ *
+ * The ConnectionEvaluator class is responsible for managing and evaluating conditions
+ * that determine when deferred connections should be executed. It provides mechanisms
+ * to evaluate connections based on specific criteria. This class is
+ * used in conjunction with the Signal class to establish and control connections.
+ */
+class ConnectionEvaluator
+{
+
+public:
+    ConnectionEvaluator() = default;
+
+    ConnectionEvaluator(const ConnectionEvaluator &) noexcept = default;
+
+    ConnectionEvaluator &operator=(const ConnectionEvaluator &) noexcept = default;
+
+    ConnectionEvaluator(ConnectionEvaluator &&other) noexcept = delete;
+
+    ConnectionEvaluator &operator=(ConnectionEvaluator &&other) noexcept = delete;
+
+    /**
+     * @brief Evaluate and execute deferred connections.
+     *
+     * This function is responsible for evaluating and executing deferred connections.
+     * It locks the `connectionsMutex` to ensure thread safety while accessing the list
+     * of deferred connections. It then copies the list of connections to avoid
+     * interference with reentrant emissions. Finally, it iterates through the copied
+     * connections and executes each one by calling the associated function object.
+     */
+    void evaluateDeferredConnections()
```

Comment:
Overall I think the approach of simply copying (or moving out of) the connections is quite cool, as it allows to quickly release the mutex again.

However, this approach conflicts with deleting connections.
When the following happens asynchronously

- emit a deferred signal
- disconnect the connection
- evaluateDeferredConnections

The connection should either be evaluated before the `disconnect` call returns, or it should not be evaluated, if the `disconnect` call returns before `evaluateDeferredConnections` is called. As it's highly likely that the connection includes dangling pointers after the call to `disconnect` the connection.

So if an evaluation is in progress the call to `disconnect` cannot return until the evaluation has completely finished.

If you still want to keep the approach of simply moving the connections over to release the mutex early, you could add a second mutex for "disconnect", which would be held longer, whilst the mutex for adding new connections could already be released.
However, this double-locking may be less performant in the end, depending on how many connections are queued, as we now need to lock two mutexes when evaluating and when disconnecting.

Reply:
I really have a doubt about this, how i am thinking about is, so follow the order 
-> emit a deferred signal
-> disconnect the connection 
-> evaluateDeferredConnections

disconnecting the signal after the emit or before the evaluation wouldn't be have any issue, as evaluation depends on the list of connections in the evaluator which has already queued up through the call to emit. Call to disconnect removes the connection from the `m_connections` of the Signal class, which is fine. 

I can only think about the issue when evaluation or disconnect either be used in mulithreading context. Or i might be thinking wrong?

I am adding the test case for this, to look up.

Reply:
Like we discussed with @seanharmer and @lemirep , both approaches are feasible.
Removing queued invocations on disconnect is really a lot better to protect from dangling references.
But as Sean noted, it might just be something that needs to be expected with a new paradigm.

So if you're not sure what to do yet, feel free to explore your initial implementation a bit further (i.e. still evaluating queued invocations even after disconnect).
I'd just like to see the implications of this in use in an actual application.
Basically: How easy is it to foot-gun yourself with it?